File: src/Tool/texto_busca_helper.py
"""Normalizacao de texto para buscas sem diferenciar maiusculas e acentos."""
from __future__ import annotations

import unicodedata


def normalizar_texto_busca(texto: str) -> str:
    """Converte para minusculas e remove acentos para comparacao de busca."""
    if not texto:
        return ""

    sem_acento = "".join(
        caractere
        for caractere in unicodedata.normalize("NFD", texto.lower())
        if unicodedata.category(caractere) != "Mn"
    )
    return sem_acento.strip()


def texto_contem_busca(termo: str, *textos: str) -> bool:
    """Verifica se o termo aparece em algum texto, ignorando caixa e acentos."""
    chave = normalizar_texto_busca(termo)
    if not chave:
        return False

    for texto in textos:
        if chave in normalizar_texto_busca(texto):
            return True
    return False

File: mobile/financeiro_app/lib/util/texto_busca_helper.dart
/// Normalizacao de texto para buscas sem diferenciar maiusculas e acentos.
String normalizarTextoBusca(String texto) {
  if (texto.isEmpty) return '';
  final minusculo = texto.toLowerCase();
  final buffer = StringBuffer();
  for (final rune in minusculo.runes) {
    final caractere = String.fromCharCode(rune);
    const mapaAcentos = {
      'á': 'a', 'à': 'a', 'ã': 'a', 'â': 'a', 'ä': 'a',
      'é': 'e', 'è': 'e', 'ê': 'e', 'ë': 'e',
      'í': 'i', 'ì': 'i', 'î': 'i', 'ï': 'i',
      'ó': 'o', 'ò': 'o', 'õ': 'o', 'ô': 'o', 'ö': 'o',
      'ú': 'u', 'ù': 'u', 'û': 'u', 'ü': 'u',
      'ç': 'c', 'ñ': 'n',
    };
    buffer.write(mapaAcentos[caractere] ?? caractere);
  }
  return buffer.toString().trim();
}

bool textoContemBusca(String termo, List<String> textos) {
  final chave = normalizarTextoBusca(termo);
  if (chave.isEmpty) return false;
  for (final texto in textos) {
    if (normalizarTextoBusca(texto).contains(chave)) return true;
  }
  return false;
}
